In [1]:
import numpy as np
import pandas as pd
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest, proportion_effectsize
from statsmodels.stats.power import NormalIndPower

np.random.seed(42)

In [3]:
expected =np.array([100,100,100,100])
observed =np.array([120,85,110,85])

dof = len(observed) - 1
print(f"Degrees of freedom: {dof}")
print(f"Chi-squared: {st.chisquare(observed, f_exp=expected)}")

Degrees of freedom: 3
Chi-squared: Power_divergenceResult(statistic=np.float64(9.5), pvalue=np.float64(0.023331360430831553))


degree of freedom: 3

p-value: 0.02

stats: 9.5

since alpha 0.02 is smaller than 0.05, we can conclude that character choice of players are not equally proportioned, so we reject H0.

Warrior class are highly over-chosen with 20 higher than expected. Mages and Healers are the least chosen with both 15 down from expected count. Since our X squared is 
9.5, we can say that rogue class is over-chosen than expected too(10 higher) 

In [5]:
contingency_array = np.array([
    [90, 60, 30],
    [110, 140, 170]
])
contingency_df = pd.DataFrame(
    contingency_array,
    columns=['Free', 'Basic', 'Premium'],
    index=['Churned', 'Retained']
)

contingency_df

,Free,Basic,Premium
Churned,90,60,30
Retained,110,140,170


In [7]:
st.chi2_contingency(contingency_df)

Chi2ContingencyResult(statistic=np.float64(42.85714285714286), pvalue=np.float64(4.939576018831198e-10), dof=2, expected_freq=array([[ 60.,  60.,  60.],
       [140., 140., 140.]]))

χ² ~ 42

p-value ~ 4.9 x 10^(-10) (nearly zero)

dof = 2

expected frequencies:
  
    60  60  60
    140 140 140

by looking the stats, we can definitely say that, we are able to reject null hypothesis because p value is nearly zero and this indicates high significance. Free tier members are less likely to retain, and premium members are less likely to churn their subscription.

In [8]:
def cramers_v(chi2, n, min_dim):
    """
    Returns Cramér's V.

    Parameters
    ----------
    chi2 : float
        Chi-square statistic.
    n : int
        Total number of observations.
    min_dim : int
        min(rows, cols) - 1 of the contingency table.
    """
    return np.sqrt(chi2 / (n * min_dim))

# Values from the previous task
chi2_stat = 42.85714285714286
total_n = 600
# Rows = 2 (Churned, Retained), Cols = 3 (Free, Basic, Premium)
min_dimension = min(2, 3) - 1  # 1

# Calculate Cramér's V
v_statistic = cramers_v(chi2_stat, total_n, min_dimension)

v_statistic

np.float64(0.2672612419124244)

Cramér's V is: 0.267

Using Cohen's benchmarks, we can say that relationship between tier and churn status is nearly medium. This means, there is a moderate chance that premium users will retain and free users will churn.